ksqldb-engine/src/main/java/io/confluent/ksql/services/SandboxedSchemaRegistryClient.java

```diff
@@ -36,21 +43,193 @@
 
   static SchemaRegistryClient createProxy(final SchemaRegistryClient delegate) {
     Objects.requireNonNull(delegate, "delegate");
-
-    return LimitedProxyBuilder.forClass(SchemaRegistryClient.class)
-        .swallow("register", anyParams(), 123)
-        .swallow("getId", anyParams(), 123)
-        .forward("getAllSubjects", methodParams(), delegate)
-        .forward("getSchemaById", methodParams(int.class), delegate)
-        .forward("getLatestSchemaMetadata", methodParams(String.class), delegate)
-        .forward("getSchemaBySubjectAndId", methodParams(String.class, int.class), delegate)
-        .forward("testCompatibility",
-            methodParams(String.class, ParsedSchema.class), delegate)
-        .swallow("deleteSubject", methodParams(String.class), Collections.emptyList())
-        .forward("getVersion", methodParams(String.class, ParsedSchema.class), delegate)
-        .build();
+    return new SandboxSchemaRegistryCache(delegate);
   }
 
   private SandboxedSchemaRegistryClient() {
   }
+
+  private static final class SandboxSchemaRegistryCache implements SchemaRegistryClient {
+    private final SchemaRegistryClient delegate;
+
+    private int nextSchemaId = Integer.MAX_VALUE;
+    private final Map<String, ParsedSchema> subjectCache = new HashMap<>();
+    private final Map<String, Integer> subjectToId = new HashMap<>();
+    private final Map<Integer, ParsedSchema> idCache = new HashMap<>();
+
+    private SandboxSchemaRegistryCache(final SchemaRegistryClient delegate) {
+      this.delegate = delegate;
+    }
+
+    @Override
+    public Optional<ParsedSchema> parseSchema(
+        final String schemaType,
+        final String schemaString,
+        final List<SchemaReference> references) {
+      throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int register(final String subject, final ParsedSchema parsedSchema) {
+      if (subjectCache.containsKey(subject)) {
+        throw new IllegalStateException("Subject '" + subject + "' already in use.");
```

Comment:
Why does this throw instead of returning existing id for the subject which is the behavior of schema registry?

Reply:
Should it not depend on the compatibility rules if updating the schema actually works? Not sure to what extend we need to fully mimic SR logic? -- I tired to keep it simple in the hope it would be sufficient. Not totally sure to be fair. Thoughts?

Reply:
This will make the following fail because they share same subject?
`CREATE test_stream_1 (id int key, name varchar) with (kafka_topic='some_topic', partitions=1, format='avro')`
`CREATE test_stream_2 (id int key, name varchar, age int) with (kafka_topic='some_topic', partitions=1, format='avro')`

Reply:
I think we should err on the side of false positives rather than false negatives. It'd be bad to have statements that actually work fail validation because then we might be introducing a regression (if a customer uses such statements in their workflows). In contrast, a false positive means we let some statements through validation that then fail in execution. Not great since multi-statements requests might have been partially executed, but also doesn't prevent any existing workflows.